src/Compilers/CSharp/Portable/Parser/SlidingTextWindow.cs

```diff
@@ -34,207 +44,177 @@ internal sealed class SlidingTextWindow : IDisposable
-            }
-            else
-            {
-                // we need to reread text buffer
-                int amountToRead = Math.Min(_text.Length, position + _characterWindow.Length) - position;
-                amountToRead = Math.Max(amountToRead, 0);
-                if (amountToRead > 0)
-                {
-                    _text.CopyTo(position, _characterWindow, 0, amountToRead);
-                }
-
-                _lexemeStart = 0;
-                _offset = 0;
-                _basis = position;
-                _characterWindowCount = amountToRead;
-            }
-        }
-
-        private bool MoreChars()
-        {
-            if (_offset >= _characterWindowCount)
-            {
-                if (this.Position >= _textEnd)
-                {
-                    return false;
-                }
-
-                // if lexeme scanning is sufficiently into the char buffer, 
-                // then refocus the window onto the lexeme
-                if (_lexemeStart > (_characterWindowCount / 4))
-                {
-                    Array.Copy(_characterWindow,
-                        _lexemeStart,
-                        _characterWindow,
-                        0,
-                        _characterWindowCount - _lexemeStart);
-                    _characterWindowCount -= _lexemeStart;
-                    _offset -= _lexemeStart;
-                    _basis += _lexemeStart;
-                    _lexemeStart = 0;
-                }
-
-                if (_characterWindowCount >= _characterWindow.Length)
-                {
-                    // grow char array, since we need more contiguous space
-                    char[] oldWindow = _characterWindow;
-                    char[] newWindow = new char[_characterWindow.Length * 2];
-                    Array.Copy(oldWindow, 0, newWindow, 0, _characterWindowCount);
-                    s_windowPool.ForgetTrackedObject(oldWindow, newWindow);
-                    _characterWindow = newWindow;
-                }
+            // Move us to that position.
+            _positionInText = Math.Min(position, _textEnd);
 
-                int amountToRead = Math.Min(_textEnd - (_basis + _characterWindowCount),
-                    _characterWindow.Length - _characterWindowCount);
-                _text.CopyTo(_basis + _characterWindowCount,
-                    _characterWindow,
-                    _characterWindowCount,
-                    amountToRead);
-                _characterWindowCount += amountToRead;
-                return amountToRead > 0;
-            }
+            // if position is within already read character range then just use what we have
+            if (PositionIsWithinWindow(position))
+                return;
 
-            return true;
+            // Otherwise, ensure that the character window contains this position so we can read characters directly
+            // from there.
+            this.ReadChunkAt(position);
```

Comment:
logic is now so much simpler.  especially not needing to figure out how to ensure lexeme positioning is still accurate.